### ISSUE TITLE
AvroProducer fails silently when not setting `"security.protocol": "ssl"`

### ISSUE THREAD

**Issue by ruurtjan** [confluentinc/confluent-kafka-python#741]
Description
===========

Today, we forgot to set the security protocol to ssl. This resulted in flush taking a couple of minutes, then proceeding without storing anything to Kafka. The schemas got persisted to the schema registry though.

Expected behaviour is some error message, preferably indicating that you forgot to specify the security protocol.

Output with debug on:
```
%7|1576156120.005|BROKER|rdkafka#producer-1| [thrd:app]: [...broker1...]:9094/bootstrap: Added new broker with NodeId -1
%7|1576156120.005|BRKMAIN|rdkafka#producer-1| [thrd::0/internal]: :0/internal: Enter main broker thread
%7|1576156120.005|BRKMAIN|rdkafka#producer-1| [thrd:[...broker1...].]: [...broker1...]:9094/bootstrap: Enter main broker thread
%7|1576156120.005|BROKER|rdkafka#producer-1| [thrd:app]: [...broker2...]:9094/bootstrap: Added new broker with NodeId -1
%7|1576156120.005|BROKER|rdkafka#producer-1| [thrd:app]: [...broker3...]:9094/bootstrap: Added new broker with NodeId -1
%7|1576156120.005|BRKMAIN|rdkafka#producer-1| [thrd:[...broker2...].]: [...broker2...]:9094/bootstrap: Enter main broker thread
%7|1576156120.005|CONNECT|rdkafka#producer-1| [thrd:app]: [...broker3...]:9094/bootstrap: Selected for cluster connection: bootstrap servers added (broker has 0 connection attempt(s))
%7|1576156120.005|INIT|rdkafka#producer-1| [thrd:app]: librdkafka v1.2.0 (0x10200ff) rdkafka#producer-1 initialized (builtin.features gzip,snappy,ssl,sasl,regex,lz4,sasl_plain,sasl_scram,plugins,zstd,sasl_oauthbearer, GCC GXX PKGCONFIG INSTALL GNULD LDS LIBDL PLUGINS ZLIB SSL ZSTD HDRHISTOGRAM SNAPPY SOCKEM SASL_SCRAM SASL_OAUTHBEARER CRC32C_HW, debug 0x46)
%7|1576156120.005|BRKMAIN|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Enter main broker thread
%7|1576156120.005|CONNECT|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Received CONNECT op
%7|1576156120.005|STATE|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Broker changed state INIT -> TRY_CONNECT
%7|1576156120.005|CONNECT|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: broker in state TRY_CONNECT connecting
%7|1576156120.005|STATE|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Broker changed state TRY_CONNECT -> CONNECT
get_producer
produce
%7|1576156120.038|CONNECT|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Connecting to ipv4#10.212.4.49:9094 (plaintext) with socket 11
%7|1576156120.074|CONNECT|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Connected to ipv4#10.212.4.49:9094
%7|1576156120.074|CONNECTED|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Connected (#1)
%7|1576156120.074|FEATURE|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Updated enabled protocol features +ApiVersion to ApiVersion
%7|1576156120.074|STATE|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Broker changed state CONNECT -> APIVERSION_QUERY
%7|1576156120.104|BROKERFAIL|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: failed: err: Local: Broker transport failure: (errno: Connection reset by peer)
%7|1576156120.104|FEATURE|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Updated enabled protocol features -ApiVersion to 
%7|1576156120.104|STATE|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Broker changed state APIVERSION_QUERY -> DOWN
%7|1576156120.104|BROKERFAIL|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: failed: err: Local: Broker transport failure: (errno: Connection reset by peer)
%7|1576156120.104|FAIL|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: ApiVersionRequest failed: Local: Broker transport failure: probably due to old broker version (after 0ms in state DOWN)
%7|1576156120.250|TOPIC|rdkafka#producer-1| [thrd:app]: New local topic: ruurtjanTest
%7|1576156120.250|TOPPARNEW|rdkafka#producer-1| [thrd:app]: NEW ruurtjanTest [-1] 0x2772ed0 (at rd_kafka_topic_new0:393)
flush
%7|1576156121.005|NOINFO|rdkafka#producer-1| [thrd:main]: Topic ruurtjanTest metadata information unknown
%7|1576156121.005|NOINFO|rdkafka#producer-1| [thrd:main]: Topic ruurtjanTest partition count is zero: should refresh metadata
%7|1576156121.005|CONNECT|rdkafka#producer-1| [thrd:main]: [...broker1...]:9094/bootstrap: Selected for cluster connection: refresh unavailable topics (broker has 0 connection attempt(s))
%7|1576156121.005|CONNECT|rdkafka#producer-1| [thrd:main]: Not selecting any broker for cluster connection: still suppressed for 49ms: refresh unavailable topics
%7|1576156121.005|CONNECT|rdkafka#producer-1| [thrd:main]: Not selecting any broker for cluster connection: still suppressed for 49ms: no cluster connection
%7|1576156121.005|CONNECT|rdkafka#producer-1| [thrd:[...broker1...].]: [...broker1...]:9094/bootstrap: Received CONNECT op
%7|1576156121.005|STATE|rdkafka#producer-1| [thrd:[...broker1...].]: [...broker1...]:9094/bootstrap: Broker changed state INIT -> TRY_CONNECT
%7|1576156121.005|CONNECT|rdkafka#producer-1| [thrd:[...broker1...].]: [...broker1...]:9094/bootstrap: broker in state TRY_CONNECT connecting
%7|1576156121.005|STATE|rdkafka#producer-1| [thrd:[...broker1...].]: [...broker1...]:9094/bootstrap: Broker changed state TRY_CONNECT -> CONNECT
%7|1576156121.036|CONNECT|rdkafka#producer-1| [thrd:[...broker1...].]: [...broker1...]:9094/bootstrap: Connecting to ipv4#10.212.11.81:9094 (plaintext) with socket 11
%7|1576156121.067|CONNECT|rdkafka#producer-1| [thrd:[...broker1...].]: [...broker1...]:9094/bootstrap: Connected to ipv4#10.212.11.81:9094
%7|1576156121.067|CONNECTED|rdkafka#producer-1| [thrd:[...broker1...].]: [...broker1...]:9094/bootstrap: Connected (#1)
%7|1576156121.067|FEATURE|rdkafka#producer-1| [thrd:[...broker1...].]: [...broker1...]:9094/bootstrap: Updated enabled protocol features +ApiVersion to ApiVersion
%7|1576156121.067|STATE|rdkafka#producer-1| [thrd:[...broker1...].]: [...broker1...]:9094/bootstrap: Broker changed state CONNECT -> APIVERSION_QUERY
%7|1576156121.104|STATE|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Broker changed state DOWN -> INIT
%7|1576156121.118|BROKERFAIL|rdkafka#producer-1| [thrd:[...broker1...].]: [...broker1...]:9094/bootstrap: failed: err: Local: Broker transport failure: (errno: Connection reset by peer)
%7|1576156121.118|FEATURE|rdkafka#producer-1| [thrd:[...broker1...].]: [...broker1...]:9094/bootstrap: Updated enabled protocol features -ApiVersion to 
%7|1576156121.118|STATE|rdkafka#producer-1| [thrd:[...broker1...].]: [...broker1...]:9094/bootstrap: Broker changed state APIVERSION_QUERY -> DOWN
%7|1576156121.118|BROKERFAIL|rdkafka#producer-1| [thrd:[...broker1...].]: [...broker1...]:9094/bootstrap: failed: err: Local: Broker transport failure: (errno: Connection reset by peer)
%7|1576156121.118|FAIL|rdkafka#producer-1| [thrd:[...broker1...].]: [...broker1...]:9094/bootstrap: ApiVersionRequest failed: Local: Broker transport failure: probably due to old broker version (after 0ms in state DOWN)
%7|1576156121.118|CONNECT|rdkafka#producer-1| [thrd:[...broker1...].]: [...broker2...]:9094/bootstrap: Selected for cluster connection: broker down (broker has 0 connection attempt(s))
%7|1576156121.118|CONNECT|rdkafka#producer-1| [thrd:[...broker1...].]: Not selecting any broker for cluster connection: still suppressed for 49ms: broker down
%7|1576156121.118|CONNECT|rdkafka#producer-1| [thrd:[...broker2...].]: [...broker2...]:9094/bootstrap: Received CONNECT op
%7|1576156121.118|STATE|rdkafka#producer-1| [thrd:[...broker2...].]: [...broker2...]:9094/bootstrap: Broker changed state INIT -> TRY_CONNECT
%7|1576156121.118|CONNECT|rdkafka#producer-1| [thrd:[...broker2...].]: [...broker2...]:9094/bootstrap: broker in state TRY_CONNECT connecting
%7|1576156121.118|STATE|rdkafka#producer-1| [thrd:[...broker2...].]: [...broker2...]:9094/bootstrap: Broker changed state TRY_CONNECT -> CONNECT
%7|1576156121.177|CONNECT|rdkafka#producer-1| [thrd:[...broker2...].]: [...broker2...]:9094/bootstrap: Connecting to ipv4#10.212.1.221:9094 (plaintext) with socket 11
%7|1576156121.209|CONNECT|rdkafka#producer-1| [thrd:[...broker2...].]: [...broker2...]:9094/bootstrap: Connected to ipv4#10.212.1.221:9094
%7|1576156121.209|CONNECTED|rdkafka#producer-1| [thrd:[...broker2...].]: [...broker2...]:9094/bootstrap: Connected (#1)
%7|1576156121.209|FEATURE|rdkafka#producer-1| [thrd:[...broker2...].]: [...broker2...]:9094/bootstrap: Updated enabled protocol features +ApiVersion to ApiVersion
%7|1576156121.209|STATE|rdkafka#producer-1| [thrd:[...broker2...].]: [...broker2...]:9094/bootstrap: Broker changed state CONNECT -> APIVERSION_QUERY
%7|1576156121.248|BROKERFAIL|rdkafka#producer-1| [thrd:[...broker2...].]: [...broker2...]:9094/bootstrap: failed: err: Local: Broker transport failure: (errno: Connection reset by peer)
%7|1576156121.248|FEATURE|rdkafka#producer-1| [thrd:[...broker2...].]: [...broker2...]:9094/bootstrap: Updated enabled protocol features -ApiVersion to 
%7|1576156121.248|STATE|rdkafka#producer-1| [thrd:[...broker2...].]: [...broker2...]:9094/bootstrap: Broker changed state APIVERSION_QUERY -> DOWN
%7|1576156121.248|BROKERFAIL|rdkafka#producer-1| [thrd:[...broker2...].]: [...broker2...]:9094/bootstrap: failed: err: Local: Broker transport failure: (errno: Connection reset by peer)
%7|1576156121.248|FAIL|rdkafka#producer-1| [thrd:[...broker2...].]: [...broker2...]:9094/bootstrap: ApiVersionRequest failed: Local: Broker transport failure: probably due to old broker version (after 0ms in state DOWN)
%7|1576156121.248|CONNECT|rdkafka#producer-1| [thrd:[...broker2...].]: [...broker3...]:9094/bootstrap: Selected for cluster connection: broker down (broker has 1 connection attempt(s))
%7|1576156121.248|CONNECT|rdkafka#producer-1| [thrd:[...broker2...].]: Not selecting any broker for cluster connection: still suppressed for 49ms: broker down
%7|1576156121.248|CONNECT|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Received CONNECT op
%7|1576156121.248|STATE|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Broker changed state INIT -> TRY_CONNECT
%7|1576156121.248|CONNECT|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: broker in state TRY_CONNECT connecting
%7|1576156121.248|STATE|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Broker changed state TRY_CONNECT -> CONNECT
%7|1576156121.320|CONNECT|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Connecting to ipv4#10.212.4.49:9094 (plaintext) with socket 11
%7|1576156121.353|CONNECT|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Connected to ipv4#10.212.4.49:9094
%7|1576156121.353|CONNECTED|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Connected (#2)
%7|1576156121.353|FEATURE|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Updated enabled protocol features +ApiVersion to ApiVersion
%7|1576156121.353|STATE|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Broker changed state CONNECT -> APIVERSION_QUERY
%7|1576156121.382|BROKERFAIL|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: failed: err: Local: Broker transport failure: (errno: Connection reset by peer)
%7|1576156121.382|FEATURE|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Updated enabled protocol features -ApiVersion to 
%7|1576156121.382|STATE|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: Broker changed state APIVERSION_QUERY -> DOWN
%7|1576156121.382|BROKERFAIL|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: failed: err: Local: Broker transport failure: (errno: Connection reset by peer)
%7|1576156121.382|FAIL|rdkafka#producer-1| [thrd:[...broker3...].]: [...broker3...]:9094/bootstrap: ApiVersionRequest failed: Local: Broker transport failure: probably due to old broker version (after 0ms in state DOWN)
%7|1576156121.382|CONNECT|rdkafka#producer-1| [thrd:[...broker3...].]: Cluster connection already in progress: broker down
%7|1576156121.382|CONNECT|rdkafka#producer-1| [thrd:[...broker3...].]: Not selecting any broker for cluster connection: still suppressed for 49ms: broker down
```

How to reproduce
================

```
    producer = AvroProducer({
        # "security.protocol": "ssl",
        'bootstrap.servers': '...',
        'schema.registry.url': '...'
    }, default_key_schema=key_schema, default_value_schema=value_schema)

    producer.produce(topic="...",
                     value={'...': '...'},
                     key={'...': '...'})
    producer.flush()
```

Checklist
=========
Please provide the following information:

 - [x] confluent-kafka-python and librdkafka version (`confluent_kafka.version()` and `confluent_kafka.libversion()`): ('1.2.0', 16908288) and ('1.2.0', 16908543)
 - [x] Apache Kafka broker version: 2.2.1
 - [x] Client configuration: `{...}` see 'how to reproduce'
 - [x] Operating system: Ubuntu
 - [x] Provide client logs (with `'debug': '..'` as necessary)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
I believe this is a duplicate of #722. 
As a workaround, implement an error_cb() to print errors

**Comment by ruurtjan** [comment]
Ah, I think it is indeed.

That issue is closed and only proposes to update the docs. Failing to persist messages should definitely throw an error though. If calling `flush` doesn't flush, it doesn't do what it promises. The default behaviour should not be to swallow errors.

**Comment by edenhill** [comment]
flush() (without timeout) will not return until all messages in queue has either been successfully delivered or fail permanently, in the case of failure to connect to the broker the permanent failure will be ERR__MSG_TIMED_OUT.

**Comment by ruurtjan** [comment]
It must be a bug then, since the program in _How to reproduce_ finished without throwing an error and without successfully delivering the record.

**Comment by edenhill** [comment]
You dont seem to have a delivery report callback defined, so the message probably fails delivery but there is no way for the producer to tell you about it.

**Comment by ruurtjan** [comment]
Ah, I think I get it now. `error_cb()` is required for you to get any errors at all, since the API is async. I had expected `flush` to throw an error if any of the records failed to persist. Or at the very least a warning or error in the logs. Instead, the application happily hums along without any sign of an issue. All seems fine, except that nothing happens. This, as a default behaviour, does seem to me like an odd and unexpected design.

**Comment by edenhill** [comment]
The delivery report callbacks is the method to propagate per-message delivery status, and I see now that the AvroProducer example lacks one, which is not good, will fix.
In fact, I don't think the delivery report callback should even be optional.

**Comment by edenhill** [comment]
See https://github.com/confluentinc/confluent-kafka-python/pull/742/files#diff-04c6e90faac2675aa89e2176d2eec7d8R140

**Comment by ruurtjan** [comment]
Yeah, removing optionality would reduce the chances for surprises :)

**Comment by edenhill** [comment]
cc @mhowlett who is currently looking into making the Python client more accessible.

**Comment by rnpridgeon** [comment]
Closed as duplicate of #722.
